src/Covenant/Type.hs

```diff
@@ -0,0 +1,322 @@
+module Covenant.Type
+  ( AbstractTy (..),
+    Renamed (..),
+    CompT (..),
+    ValT (..),
+    BuiltinFlatT (..),
+    BuiltinNestedT (..),
+    RenameError (..),
+    renameValT,
+    renameCompT,
+    RenameM,
+    runRenameM,
+  )
+where
+
+import Control.Monad (unless)
+import Control.Monad.Except (ExceptT, MonadError (throwError), runExceptT)
+import Control.Monad.State.Strict (State, evalState, gets, modify)
+import Data.Coerce (coerce)
+import Data.Functor.Classes (Eq1 (liftEq))
+import Data.Kind (Type)
+import Data.Vector (Vector)
+import Data.Vector qualified as Vector
+import Data.Vector.NonEmpty (NonEmptyVector)
+import Data.Vector.NonEmpty qualified as NonEmpty
+import Data.Word (Word64)
+
+-- | A type abstraction, using a DeBruijn index.
+--
+-- = Important note
+--
+-- This is a /relative/ representation: @'BoundAt' 1 0@ could refer to different
+-- things at different points in the ASG. Note that only the first field is a
+-- DeBruijn index: the second indicates which type variable bound at that
+-- \'level\' we mean.
+--
+-- @since 1.0.0
+data AbstractTy = BoundAt Word64 Word64
```

Comment:
The debruijn index is actually the nesting number of specific variable right? :thinking: 
i.e. the number of scopes up till the variable is bound right?

That's the reason the following (from testsuite):
`forall a b . (a -> !b) -> [a] -> !b`
The mapping function is the following:
```
ThunkT
 . CompT 0
 . NonEmpty.consV (Abstraction (BoundAt 1 0))
 $ [Abstraction (BoundAt 1 1)]
```

i.e. the `a` in the mapping function `(a -> !b)` is `BoundAt 1 0` because it takes one jump to outer scope to get to the scope where the variable is bound at, right?